File: 2nd Year/Term 2/Data Structures & Algorithms/MS1/Queue/Pseudocodes/AddandDelete.java
package Queue.Pseudocodes;

import java.util.ArrayDeque;
import java.util.Deque;

/**
 * Add & Delete in Queue - Enqueue (add) and Dequeue (delete) inventory process (MotorPH)
 */
public class AddandDelete {

    record QueueItem(String brand, int quantity) {}

    /** Enqueue inventory process (add to back of queue). */
    static void enqueue(Deque<QueueItem> queue, String brand, int quantity) {
        queue.addLast(new QueueItem(brand, quantity));
    }

    /** Dequeue inventory process (remove from front). Returns item or null if empty. */
    static QueueItem dequeue(Deque<QueueItem> queue) {
        if (queue.isEmpty()) {
            return null;
        }
        return queue.removeFirst();
    }

    /** Display queue (front to back). */
    static void display(Deque<QueueItem> queue) {
        if (queue.isEmpty()) {
            System.out.println("  (queue empty)");
            return;
        }
        int i = 1;
        for (QueueItem item : queue) {
            System.out.println("  " + i++ + ". " + item.brand() + ": " + item.quantity() + " units");
        }
    }

    public static void main(String[] args) {
        // Queue for inventory process (FIFO - ArrayDeque: addLast = enqueue, removeFirst = dequeue)
        Deque<QueueItem> queue = new ArrayDeque<>();

        // Enqueue inventory process (add)
        enqueue(queue, "Honda", 5);
        enqueue(queue, "Yamaha", 4);
        enqueue(queue, "Suzuki", 7);

        System.out.println("Queue after enqueues:");
        display(queue);

        // Dequeue inventory process (delete front)
        QueueItem process = dequeue(queue);
        System.out.println("\nDequeued: " + process);
        System.out.println("Queue after dequeue:");
        display(queue);

        // Dequeue one more
        process = dequeue(queue);
        System.out.println("\nDequeued: " + process);
        System.out.println("Queue after dequeue:");
        display(queue);
    }
}

File: 2nd Year/Term 2/Data Structures & Algorithms/MS1/Old files (mistake)/Queue/Pseudocodes/AddandDelete.py
# Add & Delete in Queue - Enqueue (add) and Dequeue (delete) inventory process (MotorPH)


def enqueue(queue, brand, quantity):
    """Enqueue inventory process (add to back of queue)."""
    queue.append({"brand": brand, "quantity": quantity})


def dequeue(queue):
    """Dequeue inventory process (remove from front). Returns item or None if empty."""
    if not queue:
        return None
    return queue.pop(0)


def display(queue):
    """Display queue (front to back)."""
    if not queue:
        print("  (queue empty)")
        return
    for i, item in enumerate(queue):
        print(f"  {i + 1}. {item['brand']}: {item['quantity']} units")


def main():
    # Queue for inventory process (FIFO)
    queue = []

    # Enqueue inventory process (add)
    enqueue(queue, "Honda", 5)
    enqueue(queue, "Yamaha", 4)
    enqueue(queue, "Suzuki", 7)

    print("Queue after enqueues:")
    display(queue)

    # Dequeue inventory process (delete front)
    process = dequeue(queue)
    print(f"\nDequeued: {process}")
    print("Queue after dequeue:")
    display(queue)

    # Dequeue one more
    process = dequeue(queue)
    print(f"\nDequeued: {process}")
    print("Queue after dequeue:")
    display(queue)


if __name__ == "__main__":
    main()
#
